quarkus/tests/integration/src/main/java/org/keycloak/it/utils/RawKeycloakDistribution.java

```diff
@@ -154,19 +144,36 @@ public void stop() {
         shutdownOutputExecutor();
     }
 
-    private void killChildProcessesOnWindows(boolean isForced) {
-        for (ProcessHandle childProcessHandle : keycloak.children().collect(Collectors.toList())) {
-            CompletableFuture<ProcessHandle> onExit = childProcessHandle.onExit();
-            if (isForced) {
-                childProcessHandle.destroyForcibly();
+    private void destroyDescendantsOnWindows(Process parent, boolean force) {
+        if (!isWindows()) {
+            return;
+        }
+
+        CompletableFuture allProcesses = CompletableFuture.completedFuture(null);
+
+        for (ProcessHandle process : parent.descendants().collect(Collectors.toList())) {
+            if (force) {
+                process.destroyForcibly();
             } else {
-                childProcessHandle.destroy();
+                process.destroy();
             }
-            //for whatever reason windows doesnt wait for the termination,
-            // and parent process returns immediately with exitCode 1 but is not exited, leading to
-            // "failed to start the distribution" bc files that should be deleted
-            // are used by another process, so we need this here.
-            onExit.join();
+
+            allProcesses = CompletableFuture.allOf(allProcesses, process.onExit());
+        }
+
+        try {
+            allProcesses.get(DEFAULT_SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        } catch (Exception cause) {
+            throw new RuntimeException("Failed to terminate descendants processes", cause);
+        }
+
+        try {
+            // TODO: remove this. do not ask why, but on Windows we are here even though the process was previously terminated
+            // without this pause, tests re-installing dist before tests should fail
+            // looks like pausing the current thread let windows to cleanup processes?
+            // more likely it is env dependent
+            Thread.sleep(500);
```

Comment:
iirc that was exactly why I added the completablefuture with the join at the end. But tbh: if it works, it's fine for me. :)

Reply:
Not really, your implementation was not taking into account interruption. See the differences in the implementation.

Reply:
all i can tell is "worked on my machine", but agree that windows is just a beast ;)